presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java

```diff
@@ -120,69 +134,104 @@ public PlanNode visitFilter(FilterNode filter, RewriteContext<Void> context)
             TupleDomain<IcebergColumnHandle> entireColumnDomain = decomposedFilter.getTupleDomain()
                     .transform(subfield -> subfield.getPath().isEmpty() ? subfield.getRootName() : null)
                     .transform(nameToColumnHandlesMapping::get);
-            boolean hasSubfieldsInNestedStructures = decomposedFilter.getTupleDomain().getDomains()
-                    .map(map -> map.keySet().stream().anyMatch(subfield -> !subfield.getPath().isEmpty()))
-                    .orElse(false);
+
+            TableHandle handle = tableScan.getTable();
+            IcebergTableHandle tableHandle = (IcebergTableHandle) handle.getConnectorHandle();
+            IcebergAbstractMetadata metadata = (IcebergAbstractMetadata) transactionManager.get(handle.getTransaction());
+            Table icebergTable = getIcebergTable(metadata, session, tableHandle.getSchemaTableName());
+
+            // Get predicate expression on subfield
+            SubfieldExtractor subfieldExtractor = new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session);
+            Map<String, com.facebook.presto.common.type.Type> columnTypes = nameToColumnHandlesMapping.entrySet().stream()
+                    .collect(toImmutableMap(entry -> entry.getKey(), entry -> entry.getValue().getType()));
+            TupleDomain<RowExpression> subfieldTupleDomain = decomposedFilter.getTupleDomain()
+                    .transform(subfield -> subfield.getPath().isEmpty() ? null : subfield)
+                    .transform(subfield -> subfieldExtractor.toRowExpression(subfield, columnTypes.get(subfield.getRootName())));
+            RowExpression subfieldPredicate = rowExpressionService.getDomainTranslator().toPredicate(subfieldTupleDomain);
+
+            // Get predicate tuple domain on identity partition columns, which could be enforced by iceberg table itself
+            List<IcebergColumnHandle> identityPartitionColumns = getIdentityPartitionColumnHandles(icebergTable,
+                    nameToColumnHandlesMapping.values().stream().collect(Collectors.toList()));
+            TupleDomain<ColumnHandle> identityPartitionColumnPredicate = TupleDomain.withColumnDomains(
+                    Maps.filterKeys(
+                            entireColumnDomain.transform(icebergColumnHandle -> (ColumnHandle) icebergColumnHandle)
+                                    .getDomains().get(),
```

Comment:
Yes, I think it should not be empty. As far as I understand, an empty `getDomain()` means the `entireColumnDomain` should be a `TupleDomain.none()`, which implying the filter expression is alway a `FALSE`. This situation should have been handled earlier, rather than being left here.

So, if it happens in some cases, I think maybe throwing an exception would be nice, so that we can consider it as a bug and fix it. If anything I misunderstood please let me know.

Reply:
Thanks. Makes more sense. If the intention is to throw an exception maybe write in `orElseThrow` with a message indicating it is intended

Reply:
Good suggestion! Added a verification at the beginning to explicitly make sure `filterPredicate` should not equals to `FALSE`.